uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/api/v3/WorkPatternTest.php

```diff
@@ -0,0 +1,116 @@
+<?php
+
+use CRM_HRCore_Test_Fabricator_Contact as ContactFabricator;
+use CRM_Hrjobcontract_Test_Fabricator_HRJobContract as HRJobContractFabricator;
+use CRM_HRLeaveAndAbsences_BAO_WorkPattern as WorkPattern;
+use CRM_HRLeaveAndAbsences_Service_JobContract as JobContractService;
+use CRM_HRLeaveAndAbsences_Service_WorkPatternCalendar as WorkPatternCalendarService;
+use CRM_HRLeaveAndAbsences_Test_Fabricator_AbsencePeriod as AbsencePeriodFabricator;
+use CRM_HRLeaveAndAbsences_Test_Fabricator_ContactWorkPattern as ContactWorkPatternFabricator;
+use CRM_HRLeaveAndAbsences_Test_Fabricator_WorkPattern as WorkPatternFabricator;
+
+/**
+ * Class api_v3_WorkPatternTest
+ *
+ * @group headless
+ */
+class api_v3_WorkPatternTest extends BaseHeadlessTest {
+
+  use CRM_HRLeaveAndAbsences_WorkPatternHelpersTrait;
+
+  /**
+   * @expectedException CiviCRM_API3_Exception
+   * @expectedExceptionMessage Mandatory key(s) missing from params array: contact_id, period_id
+   */
+  public function testGetCalendarRequiresContactIdAndPeriodID() {
+    civicrm_api3('WorkPattern', 'getCalendar');
+  }
+
+  /**
+   * @expectedException CiviCRM_API3_Exception
+   * @expectedExceptionMessage Mandatory key(s) missing from params array: contact_id
+   */
+  public function testGetCalendarRequiresContactIdIfPeriodIDIsNotEmpty() {
+    civicrm_api3('WorkPattern', 'getCalendar', ['period_id' => 1]);
+  }
+
+  /**
+   * @expectedException CiviCRM_API3_Exception
+   * @expectedExceptionMessage Mandatory key(s) missing from params array: period_id
+   */
+  public function testGetCalendarRequiresPeriodIdIfContactIDIsNotEmpty() {
+    civicrm_api3('WorkPattern', 'getCalendar', ['contact_id' => 1]);
+  }
+
+  /**
+   * @expectedException CiviCRM_API3_Exception
+   * @expectedExceptionMessage Unable to find a CRM_HRLeaveAndAbsences_BAO_AbsencePeriod with id 99989389121.
+   */
+  public function testGetCalendarThrowsAnErrorIfThePeriodIDIsNotForAnExistentAbsencePeriod() {
+    civicrm_api3('WorkPattern', 'getCalendar', ['contact_id' => 1, 'period_id' => 99989389121]);
+  }
+
+  public function testGetCalendarShouldReturnEmptyIfTheGivenContactDoesntExists() {
+    $absencePeriod = AbsencePeriodFabricator::fabricate([
+      'start_date' => CRM_Utils_Date::processDate('2015-01-05'),
+      'end_date'   => CRM_Utils_Date::processDate('2015-01-19'),
+    ]);
+
+    $result = civicrm_api3('WorkPattern', 'getCalendar', ['contact_id' => 321, 'period_id' => $absencePeriod->id]);
```

Comment:
Yeah, I thought about that, but the thing is... to fetch the contracts we really only need the contact_id, not the whole contact and it seems to be a bit too much to fetch the contact (which takes at least 4 SQL queries), just to make sure it exists. That's why I opted for the empty array instead of an exception. What do you think?